### PR TITLE
Introduce distributed Adaptive API 

### PR DESCRIPTION
**Pull request by odp**
This PR adds distributed Adaptive API to Texar-PyTorch with the help of [AdaptDL](https://adaptdl.readthedocs.io/en/latest/). 

Specifically 

1. `tx.distributed.AdaptiveDataParallel` which is similar to `torch.nn.parallel.DistributedDataParallel`, should be used to wrap the model, optimizer and lr_scheduler at the beginning of the code. This makes them distributed, restart-safe and adaptive.

2. `tx.distributed.AdaptiveDataIterator` which mimics `tx.data.DataIterator` but provides an adaptive, distributed version of it. Using this, the iterators become restart-safe (able to reposition correctly after a scale-up or scale-down of training) and distributed (data is automatically partitioned across replicas) 

Using these APIs and few others directly from the `adaptdl.torch` API, Texar-PyTorch models can be trained on an AdaptDL cluster in elastic, distributed data-parallel fashion. `examples/bert/bert_classifier_adaptive.py` is the adaptive version of `examples/bert/bert_classifier_main.py`  which demonstrates the use of above API. It can be trained on a cluster by running `examples/bert/run_bert_adaptive.sh` after setting up a AdaptDL kubernetes cluster or microk8s environment. 

The API is compatible with [standalone](https://adaptdl.readthedocs.io/en/latest/standalone-training.html) training on a single node machine or a multinode (non-k8s) cluster. However this mode does not support elasticity, but can be used for testing.

### CODE REVIEW

**Review by hunterhector — CHANGES_REQUESTED** [review]
Sorry for delaying the review. I have some general comments here.

I will still need to try run this myself to provide more comments, but we can fix some small stuff before that.

File: examples/bert/README.md [hunk 1]
```diff
@@ -15,7 +15,7 @@ To summarize, this example showcases:
 * Building and fine-tuning on downstream tasks
 * Use of Texar `RecordData` module for data loading and processing
 * Use of Texar `Executor` module for simplified training loops and TensorBoard visualization
-* Use of [Hyperopt]((https://github.com/hyperopt/hyperopt)) library to tune hyperparameters with 
+* Use of [Hyperopt]((https://github.com/hyperopt/hyperopt)) library to tune hyperparameters with
```
[the remark this answers is not in the document]
Comment (odp) [reply examples/bert/README.md:18]: done

File: examples/bert/bert_classifier_adaptive.py [hunk 2]
```diff
@@ -0,0 +1,245 @@
...
+config_downstream = importlib.import_module(args.config_downstream)
+config_downstream = {
+    k: v for k, v in config_downstream.__dict__.items()
+    if not k.startswith('__') and k != "hyperparams"}
+
+device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+
+logging.root.setLevel(logging.INFO)
+
+adaptdl.torch.init_process_group("nccl" if
```
[the remark this answers is not in the document]
Comment (odp) [reply examples/bert/bert_classifier_adaptive.py:71]: done. Added links to AdaptDL documentation as well.

File: texar/torch/distributed/parallel.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+# limitations under the License.
+
+import torch
+import adaptdl.torch
+from typing import Optional
+from torch.optim.optimizer import Optimizer
+from torch.optim.lr_scheduler import _LRScheduler as LRScheduler
+from texar.torch.run.executor_utils import Instance
+
+class AdaptiveDataParallel(adaptdl.torch.AdaptiveDataParallel):
```
[the remark this answers is not in the document]
Comment (odp) [reply texar/torch/distributed/parallel.py:23]: High-level classes like the `BertClassifier` add their own members like `num_classes` or `is_binary` which are used inside the model training code. `AdaptiveDataParallel` strictly returns a `torch.nn.Module`. This results in a failure when these extra members are accessed from the model training code.

File: docker/Dockerfile [hunk 4]
```diff
@@ -0,0 +1,32 @@
+# Copyright 2019 The Texar Authors. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (odp) [reply docker/Dockerfile:1]: sure

File: requirements.txt [hunk 5]
```diff
@@ -4,3 +4,4 @@ numpy >= 1.15.4
 mypy_extensions >= 0.4.1
 regex >= 2018.01.10
 sentencepiece >= 0.1.8
+adaptdl >= 0.2.4
```
[the remark this answers is not in the document]
Comment (odp) [reply requirements.txt]: Ok. Added a new group `dist`

File: examples/bert/requirements.txt [hunk 6]
```diff
@@ -1,4 +1,4 @@
 tensorflow # used for loading BERT official model checkpoint
 tensorboardX>=1.8 # used only in bert_classifier_using_executor_main.py
 hyperopt # used only in bert_with_hypertuning_main.py
-nni # used only in bert_executor_hypertuning_nni.py
\ No newline at end of file
+nni # used only in bert_executor_hypertuning_nni.py
```
[the remark this answers is not in the document]
Comment (odp) [reply examples/bert/requirements.txt:4]: Actually nni has some issues when used with the official pytorch 1.6.0 docker image. I have included the dependencies directly in the `Dockerfile`.

**Review by hunterhector — APPROVED** [review]

**Pull request merged**